src/test/pages/Contact.test.tsx

```diff
@@ -0,0 +1,228 @@
+    allFormFields.forEach((field, index) => expect(field).toHaveValue(expectedValues[index]));
+  });
+
+  const fillOutForm = async () => {
+    const { firstName, subject, message } = formFields;
+    await waitFor(() => expect(firstName()).not.toHaveValue(""));
+    const error = screen.queryByRole("alert");
+    expect(error).toBeNull();
+    await userEvent.type(subject(), "Test subject");
+    await userEvent.type(message(), "Test message");
+    await clickButton("Submit");
+    expect(contactFormSubmitSpy).toHaveBeenCalledWith({
+      firstName: "Test",
+      lastName: "Admin",
+      emailAddress: "test-admin@test.com",
+      subject: "Test subject",
+      message: "Test message",
+    });
+  };
+
+  it("submits with expected values when form is filled correctly, and shows success message", async () => {
+    renderContactUs();
+    await fillOutForm();
+    const successMessage = screen.getByText(/Thank you for your message/i);
+    expect(successMessage).toBeInTheDocument();
+    const error = screen.queryByRole("alert");
+    expect(error).toBeNull();
+  });
+
+  it("shows error message if form submission fails", async () => {
+    renderContactUs({ successfulRequest: false });
+    await fillOutForm();
+    const successMessage = screen.queryByText(/Thank you for your message/i);
+    expect(successMessage).toBeNull();
+    const error = screen.getByRole("alert");
+    expect(error).toBeInTheDocument();
+  });
+});
+
+describe("Contact form presets", () => {
+  let originalLocation: Location;
+
+  beforeEach(() => {
+    originalLocation = window.location;
+    Object.defineProperty(window, "location", {
+      writable: true,
+      value: { ...originalLocation },
+    });
+  });
+
+  afterEach(() => {
+    Object.defineProperty(window, "location", {
+      writable: true,
+      value: originalLocation,
+    });
+  });
+
+  const presetTestCases = [
+    {
+      preset: "contentProblem",
+      expectedSubject: "Content problem",
+      expectedMessage: "Please describe the problem here",
+    },
+    {
+      preset: "teacherRequest",
+      expectedSubject: "Teacher Account Request",
+      expectedMessage: "Please could you convert my Isaac account into a teacher account",
+    },
+    {
+      preset: "accountDeletion",
+      expectedSubject: "Account Deletion Request",
+      expectedMessage: "Please could you delete my Isaac Computer Science account",
+    },
+  ];
+  it.each(presetTestCases)(
+    "shows $expectedSubject in subject if this information is provided in the URL, and pre-fills message",
+    async ({ preset, expectedSubject, expectedMessage }) => {
+      renderContactUs();
+      setLocation(`?preset=${preset}`);
+      const { firstName, subject, message } = formFields;
+      await waitFor(() => expect(firstName()).not.toHaveValue(""));
+      expect(subject()).toHaveValue(expectedSubject);
+      if (preset === "contentProblem") {
```

Comment:
Having if/elses inside a test case is normally a sign that the test is doing too much, or being too clever (and therefore, harder to understand). Is there a way this can be refactored to express your intention in a clearer way?

Reply:
I was trying to combine the tests because they were for the most part very repetitive with the exception of slightly different handling of message value vs placeholder text. I've separated out the placeholder one into it's only test, to avoid the if/else. Similarly in other tests where I took the same approach.